kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,238 @@
+
+impl<const S: usize> BlockIndex<S> {
+    /// Returns the index that contains the address.
+    pub fn new_containing(address: u64) -> Self {
+        Self((address / S as u64) as u32)
+    }
+
+    /// Returns the index starting at the given address, if any.
+    pub fn new_starting_at(address: u64) -> Option<Self> {
+        if address % S as u64 == 0 {
+            Some(Self::new_containing(address))
+        } else {
+            None
+        }
+    }
+}
+
+impl<const S: usize> From<BlockIndex<S>> for u64 {
+    fn from(index: BlockIndex<S>) -> Self {
+        index.0 as u64 * S as u64
+    }
+}
+
+impl<const S: usize> Add<u32> for BlockIndex<S> {
+    type Output = Self;
+    fn add(self, other: u32) -> Self {
+        BlockIndex(self.0 + other)
+    }
+}
+
+/// Writeable persistent block storage device.
+///
+/// The device is formed from equally-sized storage blocks,
+/// which are arranged one after another, without gaps or overlaps,
+/// to form a linear storage of bytes.
+///
+/// There device is split into blocks in two ways, into:
+/// - prepare blocks, which are the smallest unit of space
+/// that can be prepared for writing (see `BlockStorage::prepare_write`)
+/// - write blocks, which are the smallest unit of space that can be written
+///
+/// Every byte on the device belongs to exactly one prepare block,
+/// and to exactly one write block at the same time.
+///
+/// `W`: The size of a write block in bytes.
+/// `P`: The size of a prepare block in bytes.
+pub trait BlockStorage<const W: usize, const P: usize> {
+    /// Read data from a block, and into the buffer.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `Client::read_complete` once.
+    fn read(
+        &self,
+        region: &BlockIndex<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Write data from a buffer to storage.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// starting at the chosen `start_block`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT prepare the block for writing first.
+    /// The user of this function SHOULD ensure that a block is prepared
+    /// before calling `write` (see `prepare_write`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it SHOULD NOT be written again until it's prepared
+    /// as part of a prepare block.
+    /// Multiple writes to the same block are possible in this trait,
+    /// but the result is dependent on the nature of the underlying hardware,
+    /// and this interface does not provide a way to discover that.
```

Comment:
I am wary of this SHOULD NOT. It means that someone can use an instance of this trait and make implementation-specific or hardware-specific assumptions, which then break on another implementation. This should either not be general (distinguish write models), or a MUST.

Reply:
Sure. I wanted to accommodate NOR flash devices (they allow multiple writes without erasing), but it's probably better to not touch it until there's an actual need.

Reply:
I believe some NAND devices do as well (although fewer writes).

Reply:
We should accommodate multiple writes without an erase. It's a pretty common use case, one that [TicKV](https://github.com/tock/tock/tree/master/libraries/tickv) and [Yaffs1](https://yaffs.net/documents/how-yaffs-works) both rely on for example

Reply:
I want to address this in a future revision. Currently we managed to work out a compromise based on the agreement that the targeted devices are basic, and the user does not need to know about how many writes they can receive.
Besides, the current version Works For Me™ ;) That is, it's already covering some useful ground.